### ISSUE TITLE
Reproducibility in YOLOX

### ISSUE THREAD

**Issue by Ibraheem951** [Megvii-BaseDetection/YOLOX#1612]
Presently, YOLOX code is not reproducible and outputs are different everytime. One of the reasons is that seed is randomly set for workers in yolox/data/dataloading.py file. Is there a way to make the pipeline reproducible?

**Comment by FateScript** [comment]
Could you provide your command?

If you want a reproducible code,  the following [gist](https://gist.github.com/ihoromi4/b681a9088f348942b01711f251e5f964) might help you.
```python3
def seed_everything(seed: int):
    import random, os
    import numpy as np
    import torch
    
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = True
    
seed_everything(42)
```

**Comment by dangnh0611** [comment]
Same problem here, i can't reproduce my results with exactly same configurations (exp file) with `seed` used.
What I have tried:
1. Use seed in exp file
2. Overwrite exp's seed in command line: `python -m yolox.tools.train -f exps/example/custom/my_custom_exp.py -d 1 -b 64 -o -c pretrains/yolox_nano.pth seed 42`
3. Try your `seed_everything()` like above in `tools/train.py` 
```python
if exp.seed is not None:
        exp.seed = int(exp.seed)
        seed_everything(exp.seed)
        warnings.warn(
            "You have chosen to seed training. This will turn on the CUDNN deterministic setting, "
            "which can slow down your training considerably! You may see unexpected behavior "
            "when restarting from checkpoints."
        )
```
4. Change worker init function in `yolox/data/dataloading.py`, from
```python
def _worker_init_reset_seed(worker_id):
    seed = uuid.uuid4().int % 2**32
    random.seed(seed)
    torch.set_rng_state(torch.manual_seed(seed).get_state())
    np.random.seed(seed)
```

to

```python
def worker_init_reset_seed(seed):
    def _seed_all(worker_id):
        from yolox.utils import seed_everything
        worker_seed = worker_id + seed        
        seed_everything(worker_seed)
        print(f'DO WORKER INIT RESET SEED WITH SEED = {worker_seed}, WORKER_ID = {worker_id}')
    return _seed_all
```

Any workaround ? Many thanks !

**Comment by wiedermaschinisch** [comment]
I'm facing the same problem. Have you since found a way to make the training reproducable 
@daquexian  @Ibraheem951?

**Comment by ZhangLin-PKU** [comment]
@pasin-k That's the last part I missed! Now I can reproduce th results.

For anyone facing this problem, what I have done is:

1. Use seed in exp file: self.seed = 42
2. Add set_deterministic_training in tools/train.py, so the code changes from
````bash
    if exp.seed is not None:
        random.seed(exp.seed)
        torch.manual_seed(exp.seed)
        cudnn.deterministic = True

        warnings.warn(
            "You have chosen to seed training. This will turn on the CUDNN deterministic setting, "
            "which can slow down your training considerably! You may see unexpected behavior "
            "when restarting from checkpoints.")

    # set environment variables for distributed training
    configure_nccl()
    configure_omp()
    cudnn.benchmark = True
````
to
````bash
    if exp.seed is not None:
        # random.seed(exp.seed)
        # torch.manual_seed(exp.seed)
        # cudnn.deterministic = True
        set_deterministic_training(exp.seed)

        warnings.warn(
            "You have chosen to seed training. This will turn on the CUDNN deterministic setting, "
            "which can slow down your training considerably! You may see unexpected behavior "
            "when restarting from checkpoints.")
    else:
        cudnn.benchmark = True

    # set environment variables for distributed training
    configure_nccl()
    configure_omp()
    # cudnn.benchmark = True
````
My set_deterministic_training is a little different from seed_everything
````bash
def set_deterministic_training(seed: int = 42) -> None:
    """
    Set seeds and deterministic settings for fully reproducible training.
    
    Note: Enabling full determinism may reduce training performance due to
    disabled CUDA optimizations.
    
    Args:
        seed: Random seed for reproducibility.
    """
    logger.info(f"Setting deterministic training with seed: {seed}")

    # Set Python random seed
    random.seed(seed)

    # Set NumPy random seed
    np.random.seed(seed)

    # Set PyTorch random seeds
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)  # For multi-GPU setups

    # Set deterministic behavior for CuDNN
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

    # Set deterministic algorithms (PyTorch 1.12+)
    try:
        torch.use_deterministic_algorithms(True)
        logger.info("Using deterministic algorithms")
    except AttributeError:
        logger.warning(
            "torch.use_deterministic_algorithms not available in this PyTorch version"
        )

    # Set environment variables for additional determinism
    os.environ["PYTHONHASHSEED"] = str(seed)
    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"  # For deterministic CUDA

    logger.info("Deterministic training setup complete")
````
3. change worker_init_reset_seed function in yolox/data/dataloading.py to 
````bash
def worker_init_reset_seed(worker_id):
    """Initialize each worker with a different random seed."""
    # seed = uuid.uuid4().int % 2**32
    seed = 42 + worker_id
    random.seed(seed)
    torch.set_rng_state(torch.manual_seed(seed).get_state())
    np.random.seed(seed)
````

Every step is necessary.
